File: retrofitrx/src/main/java/com/shuai/retrofitrx/net/Interceptor/AuthParamsInterceptor.java
package com.shuai.retrofitrx.net.Interceptor;

import com.shuai.retrofitrx.config.NetConfig;
import com.shuai.retrofitrx.config.provider.NetRequestConfigProvider;
import com.shuai.retrofitrx.utils.Util;

import java.io.IOException;
import java.util.Collections;
import java.util.Map;

import okhttp3.FormBody;
import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;


public class AuthParamsInterceptor implements Interceptor {

    @Override
    public Response intercept(Chain chain) throws IOException {
        Request originRequest = chain.request();

        // 添加公共参数
        HttpUrl.Builder newUrlBuilder = originRequest.url().newBuilder();
        Map<String, String> defaultParams = getDefaultUrlParams();
        if (defaultParams != null && !defaultParams.isEmpty()) {
            for (Map.Entry<String, String> param : defaultParams.entrySet()) {
                newUrlBuilder.addQueryParameter(param.getKey(), param.getValue());
            }
        }

        // 添加公共 Header
        Request.Builder newRequestBuilder = originRequest.newBuilder()
                .url(newUrlBuilder.build());
        Map<String, String> defaultHeaders = getDefaultHeaders();
        if (defaultHeaders != null && !defaultHeaders.isEmpty()) {
            for (Map.Entry<String, String> entry : defaultHeaders.entrySet()) {
                newRequestBuilder.addHeader(entry.getKey(), entry.getValue());
            }
        }


        // Post 请求添加统一Body
        if ("POST".equalsIgnoreCase(originRequest.method())) {
            if (originRequest.body() instanceof FormBody) {
                Map<String, String> defaultBodyParams = getDefaultBodyParams();
                if (!Util.MapIsEmpty(defaultBodyParams)) {
                    FormBody.Builder newFormBody = new FormBody.Builder();
                    FormBody oldFormBody = (FormBody) originRequest.body();
                    if (oldFormBody != null && oldFormBody.size() > 0) {
                        for (int i = 0; i < oldFormBody.size(); i++) {
                            newFormBody.addEncoded(oldFormBody.encodedName(i), oldFormBody.encodedValue(i));
                        }
                    }
                    for (Map.Entry<String, String> entry : defaultBodyParams.entrySet()) {
                        if (entry != null && entry.getKey() != null && entry.getValue() != null) {
                            newFormBody.add(entry.getKey(), entry.getValue());
                        }
                    }
                    newRequestBuilder.method(originRequest.method(), newFormBody.build());
                }
            }
        }

        Request newRequest = newRequestBuilder.build();
        return chain.proceed(newRequest);
    }

    private Map getDefaultUrlParams() {
        NetRequestConfigProvider provider = NetConfig.getConfig().getRequestConfigProvider();
        if (provider != null) {
            return provider.getParamsMap();
        }
        return Collections.EMPTY_MAP;
    }

    private Map getDefaultHeaders() {
        NetRequestConfigProvider provider = NetConfig.getConfig().getRequestConfigProvider();
        if (provider != null) {
            return provider.getHeaderMap();
        }
        return Collections.EMPTY_MAP;
    }

    private Map getDefaultBodyParams() {
        NetRequestConfigProvider provider = NetConfig.getConfig().getRequestConfigProvider();
        if (provider != null) {
            return provider.getBodyMap();
        }
        return Collections.EMPTY_MAP;
    }

}

File: retrofitrx/src/main/java/com/shuai/retrofitrx/net/interceptor/AuthParamsInterceptor.kt
package com.shuai.retrofitrx.net.interceptor

import com.shuai.retrofitrx.config.NetConfig
import com.shuai.retrofitrx.config.provider.NetRequestConfigProvider
import com.shuai.retrofitrx.utils.Util
import okhttp3.FormBody
import okhttp3.Interceptor
import okhttp3.Response
import java.io.IOException

class AuthParamsInterceptor : Interceptor {

    private var netRequestConfigProvider: NetRequestConfigProvider?

    constructor(netRequestConfigProvider: NetRequestConfigProvider?) {
        this.netRequestConfigProvider = netRequestConfigProvider
    }

    constructor() {
        netRequestConfigProvider = NetConfig.config.requestConfigProvider
    }

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val originRequest = chain.request()

        // 添加公共参数
        val newUrlBuilder = originRequest.url().newBuilder()
        val defaultParams: Map<String, String> = defaultUrlParams
        if (defaultParams.isNotEmpty()) {
            for ((key, value) in defaultParams) {
                newUrlBuilder.addQueryParameter(key, value)
            }
        }

        // 添加公共 Header
        val newRequestBuilder = originRequest.newBuilder().url(newUrlBuilder.build())
        val defaultHeaders: Map<String, String> = defaultHeaders
        if (defaultHeaders.isNotEmpty()) {
            for ((key, value) in defaultHeaders) {
                newRequestBuilder.addHeader(key, value)
            }
        }

        // Post 请求添加统一Body
        if ("POST".equals(originRequest.method(), ignoreCase = true)) {
            if (originRequest.body() is FormBody) {
                val defaultBodyParams: Map<String, String> = defaultBodyParams
                if (!Util.mapIsEmpty(defaultBodyParams)) {
                    val newFormBody = FormBody.Builder()
                    val oldFormBody = originRequest.body() as FormBody?
                    if (oldFormBody != null && oldFormBody.size() > 0) {
                        for (i in 0 until oldFormBody.size()) {
                            newFormBody.addEncoded(oldFormBody.encodedName(i), oldFormBody.encodedValue(i))
                        }
                    }
                    for (entry in defaultBodyParams.entries) {
                        newFormBody.add(entry.key, entry.value)
                    }
                    newRequestBuilder.method(originRequest.method(), newFormBody.build())
                }
            }
        }
        val newRequest = newRequestBuilder.build()
        return chain.proceed(newRequest)
    }

    private val defaultUrlParams: Map<String, String>
        get() = if (netRequestConfigProvider != null) netRequestConfigProvider!!.paramsMap else emptyMap()

    private val defaultHeaders: Map<String, String>
        get() = if (netRequestConfigProvider != null) netRequestConfigProvider!!.headerMap else emptyMap()

    private val defaultBodyParams: Map<String, String>
        get() = if (netRequestConfigProvider != null) netRequestConfigProvider!!.bodyMap else emptyMap()

}